orderer/common/cluster/deliver_test.go

```diff
@@ -180,9 +180,23 @@ func (ds *deliverServer) Deliver(stream orderer.AtomicBroadcast_DeliverServer) e
 	if err != nil {
 		panic(err)
 	}
+
+	// FAB-16233 This is meant to mitigate timeouts when
+	// seekAssertions does not receive a value
+	timeoutC := make(chan bool, 1)
+
+	go func() {
+		time.Sleep(3 * time.Minute)
+		timeoutC <- true
+	}()
+
+	select {
+	case <-timeoutC:
+		panic("timed out waiting for replicate chains to finish")
```

Comment:
why not assert.Fatalf or something that is related to the test framework?

Reply:
Does assert have a Fatal that doesn't require consuming a testing.T?

Reply:
IIRC, this deliverServer instance is running in a go routine that is not managed by go test’s testing framework. If you use it, it simply delegates to T.FailNow which can’t be used from any go routine other than the one spawned by go test.

https://golang.org/pkg/testing/#T.FailNow

```
FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test. Calling FailNow does not stop those other goroutines.
```

So, if there are any assertions like that in the deliver server, they’re also broken.

Reply:
For example, this code https://github.com/hyperledger/fabric/blob/5389a1d9696718cb7e93a20ca8b8a2cacff3360a/orderer/common/cluster/deliver_test.go#L634-L640

is calling resurect from a go routine. If anything goes wrong in resurrect that results in this

https://github.com/hyperledger/fabric/blob/5389a1d9696718cb7e93a20ca8b8a2cacff3360a/orderer/common/cluster/deliver_test.go#L250

executing, it’s probably not going to work as you’d expect.

From a glance at the tests, this pattern is all over.

Reply:
> Does assert have a Fatal that doesn't require consuming a testing.T?

We have a testing.T in the instance as a field.

> IIRC, this deliverServer instance is running in a go routine that is not managed by go test’s testing framework. If you use it, it simply delegates to T.FailNow which can’t be used from any go routine other than the one spawned by go test.

What do you mean can't? It [still fails the test](https://play.golang.org/p/i5hPuxrPwSh), just not immediately. 
She has a select so it won't timeout, but it will fail. I just think that using the test framework to fail is more elegant than panicking, that's all.

Reply:
- The code already panics a few lines up from where this is being added.
- A fatal timeout is what we desire (ie. FailNow via require instead of assert). This is not asserting on a condition as much as mitigating test issues that are resulting in a hang.
- FailNow / require will not terminate tests out of a go routine.

I feel I need to stress that the only reason this is being added to terminate these tests in 3 minutes instead of 20 when things get locked up. While the preference would be to fix the cause(s) of the hangs, the scope is too large for people that have not worked in this area of the code.